### PR TITLE
Enable streaming ingest in AIR

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

This adds the following options to DatasetConfig, which can be used to enable streaming ingest.

```
    # Whether the dataset should be streamed into memory using pipelined reads.
    # When enabled, get_dataset_shard() returns DatasetPipeline instead of Dataset.
    # The amount of memory to use is controlled by `stream_window_size`.
    # False by default for all datasets.
    use_stream_api: Optional[bool] = None

    # Configure the streaming window size in bytes. A typical value is something like
    # 20% of object store memory. If set to -1, then an infinite window size will be
    # used (similar to bulk ingest). This only has an effect if use_stream_api is set.
    # Set to 1.0 GiB by default.
    stream_window_size: Optional[float] = None

    # Whether to enable global shuffle (per pipeline window in streaming mode). Note
    # that this is an expensive all-to-all operation, and most likely you want to use
    # local shuffle instead.
    # False by default for all datasets.
    global_shuffle: Optional[bool] = None
```

TODO:
- [x] Docs
- [x] Unit tests

### CODE REVIEW

File: python/ray/air/train/integrations/huggingface/huggingface_trainer.py [hunk 1]
```diff
@@ -240,12 +240,12 @@ def trainer_init_per_worker(train_dataset, eval_dataset, **config):
     _dataset_config = {
         "train": DatasetConfig(
             required=True,
-            streamable=False,
-            _noncustomizable_fields=["streamable"],
+            use_stream_api=False,
+            _noncustomizable_fields=["use_stream_api"],
```
Comment (ericl) [inline python/ray/air/train/integrations/huggingface/huggingface_trainer.py:244]: @gjoliver what do you think about changing this to support an arbitrary validation function, like

```
_validator=DisallowUse(["use_stream_api"]),
```

File: python/ray/air/config.py [hunk 2]
```diff
@@ -149,31 +148,46 @@ class DatasetConfig:
...
     # True by default for all datasets.
     transform: Optional[bool] = None
 
-    # List of fields that the user cannot override in ``dataset_config``.
-    _noncustomizable_fields: Tuple[str] = ()
+    # Whether the dataset should be streamed into memory using pipelined reads.
+    # When enabled, get_dataset_shard() returns DatasetPipeline instead of Dataset.
+    # The amount of memory to use is controlled by `stream_window_size`.
+    # False by default for all datasets.
+    use_stream_api: Optional[bool] = None
```
Comment (clarkzinzow) [inline python/ray/air/config.py:160]: Nit: I feel like `use_stream_api` is more verbose than it needs to be, any opposition to something like `streaming` or `use_streaming`?
```suggestion
    streaming: Optional[bool] = None
```

Comment (ericl) [reply python/ray/air/config.py:160]: Hmm, my rationale here was to call out that this is changing the class returned by "get_dataset_shard()", which is changing more than just the underlying impl.

Not sure if users get this though.

File: python/ray/air/config.py [hunk 3]
```diff
@@ -149,31 +148,46 @@ class DatasetConfig:
...
+
+    # Whether to enable global shuffle (per pipeline window in streaming mode). Note
+    # that this is an expensive all-to-all operation, and most likely you want to use
+    # local shuffle instead.
+    # False by default for all datasets.
+    global_shuffle: Optional[bool] = None
+
+    # Validator function for the final config, which should raise ValueError if the
+    # config cannot be validated.
+    _validator: Optional[Callable[["DatasetConfig"], None]] = None
```
Comment (gjoliver) [inline python/ray/air/config.py:176]: I feel like you may need to support a list of validators, like:
```
_validators=[DisallowUsage("use_stream_api"), HuggingFaceCfgValidation()]
```
another option I think is to tell users to override the validate():
```
class HuggingFaceTrainer:
  def validate(...):
    super().validate(...)
    # More custom validation I'd like to add.
```

Comment (ericl) [reply python/ray/air/config.py:176]: Ok yeah, just doing it in validate() is much better, changed!

File: python/ray/air/train/integrations/huggingface/huggingface_trainer.py [hunk 4]
```diff
@@ -308,20 +301,11 @@ def _validate_trainer_init_per_worker(
...
-                f"Got {list(self.datasets.keys())}"
-            )
-        if not all(
-            key in (TRAIN_DATASET_KEY, EVALUATION_DATASET_KEY) for key in self.datasets
-        ):
-            raise KeyError(
-                f"Only '{TRAIN_DATASET_KEY}' and '{EVALUATION_DATASET_KEY}' "
-                "keys can be preset in `datasets`. "
-                f"Got {list(self.datasets.keys())}"
-            )
```
Comment (ericl) [inline python/ray/air/train/integrations/huggingface/huggingface_trainer.py:324]: All this is auto validated now based on dataset_config.

File: doc/source/ray-air/doc_code/air_ingest.py [hunk 5]
```diff
@@ -139,3 +139,71 @@ def make_ds_2():
...
+    # Manually iterate over the data 10 times (10 epochs).
+    for _ in range(10):
+        for batch in data_shard.iter_batches():
+            print("Do some training on batch", batch)
+
+    # View the stats for performance debugging.
+    print(data_shard.stats())
+
+
+my_trainer = DataParallelTrainer(
```
Comment (amogkam) [inline doc/source/ray-air/doc_code/air_ingest.py:164]: `DataParallelTrainer` is a developer API so might be confusing to users to show this as an example. For these examples could we show usage with or TorchTrainer or TFTrainer?

File: python/ray/air/config.py [hunk 6]
```diff
@@ -146,34 +145,44 @@ class DatasetConfig:
...
-    streamable: Optional[bool] = None
-
     # Whether to transform the dataset with the fitted preprocessor. This must be
     # enabled at least for the dataset that is fit.
-    # True by default for all datasets.
+    # True by default.
     transform: Optional[bool] = None
 
-    # List of fields that the user cannot override in ``dataset_config``.
-    _noncustomizable_fields: Tuple[str] = ()
```
Comment (amogkam) [inline python/ray/air/config.py:164]: why remove `_noncustomizable_fields`?

File: python/ray/air/train/data_parallel_ingest.py [hunk 7]
```diff
@@ -76,8 +84,22 @@ def get_dataset_shards(
...
+                        bytes_per_window=config.stream_window_size
+                    ).repeat()
+                    # In windowed mode, we re-apply the preprocessor on each iteration.
+                    if self.preprocessor:
+                        prep = self.preprocessor.transform_batch
+                        dataset = dataset.map_batches(prep, batch_format="pandas")
+                else:
+                    # If the window size is infinity, the preprocessor is cached and
+                    # we don't need to re-apply it each time.
+                    dataset = dataset.repeat()
```
Comment (amogkam) [inline python/ray/air/train/data_parallel_ingest.py:99]: Can this logic all be done in `preprocess_datasets`? A bit confusing right now to have preprocessing logic in two different methods depending on if streaming is True.

File: python/ray/air/train/integrations/huggingface/huggingface_trainer.py (hunk 4 above)
Comment (amogkam) [reply python/ray/air/train/integrations/huggingface/huggingface_trainer.py:324]: Nice!

File: python/ray/air/config.py (hunk 6 above)
Comment (ericl) [reply python/ray/air/config.py:164]: It's handled via trainer validate_attributes() now.

File: python/ray/air/train/data_parallel_ingest.py (hunk 7 above)
Comment (ericl) [reply python/ray/air/train/data_parallel_ingest.py:99]: Nack. I think this split is more clear for now given existing need to be compatible with DatasetSpec signature.

File: doc/source/ray-air/doc_code/air_ingest.py (hunk 5 above)
Comment (ericl) [reply doc/source/ray-air/doc_code/air_ingest.py:164]: I changed it to TorchTrainer, but decline to show to_torch/to_tf. Those APIs are too framework specific, and also a little half baked right now.

File: python/ray/air/train/data_parallel_ingest.py (hunk 7 above)
Comment (gjoliver) [reply python/ray/air/train/data_parallel_ingest.py:99]: quite a few branches here. I wonder if the users need to care about these at all?

**Review by gjoliver — APPROVED** [review]
1 nit and 1 question. but I don't think they are blocking.

File: python/ray/air/train/data_parallel_ingest.py (hunk 7 above)
Comment (ericl) [reply python/ray/air/train/data_parallel_ingest.py:99]: From user POV it should be much easier to understand: either use batch or streaming. With streaming you can also have infinite window size as an edge case.

File: python/ray/air/config.py [hunk 8]
```diff
@@ -238,27 +247,23 @@ def validated(
...
-            _noncustomizable_fields=self._noncustomizable_fields,
+            use_stream_api=self.use_stream_api
+            if other.use_stream_api is None
+            else other.use_stream_api,
+            stream_window_size=self.stream_window_size
+            if other.stream_window_size is None
+            else other.stream_window_size,
+            global_shuffle=self.global_shuffle
+            if other.global_shuffle is None
+            else other.global_shuffle,
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/air/config.py:264]: Yeah, was wondering but I guess this config is not likely to change often.

File: python/ray/air/config.py (hunk 2 above)
Comment (clarkzinzow) [reply python/ray/air/config.py:160]: Ah, got it, thanks for adding that to the comment above!

File: python/ray/air/config.py [hunk 9]
```diff
@@ -146,34 +145,44 @@ class DatasetConfig:
...
+
+    # Configure the streaming window size in bytes. A good value is something like
+    # 20% of object store memory. If set to -1, then an infinite window size will be
+    # used (similar to bulk ingest). This only has an effect if use_stream_api is set.
+    # Set to 1.0 GiB by default.
+    stream_window_size: Optional[float] = None
+
+    # Whether to enable global shuffle (per pipeline window in streaming mode). Note
+    # that this is an expensive all-to-all operation, and most likely you want to use
+    # local shuffle instead.
```
Comment (clarkzinzow) [inline python/ray/air/config.py:170]: We should create a shuffling feature guide that we can link to, one that expands on this FAQ: https://docs.ray.io/en/master/data/faq.html#when-should-i-use-global-per-epoch-shuffling

File: python/ray/air/config.py [hunk 10]
```diff
@@ -146,34 +145,44 @@ class DatasetConfig:
...
         """Return a copy of this config with all default values filled in."""
         return DatasetConfig(
             fit=self.fit or False,
             split=self.split or False,
             required=self.required or False,
-            streamable=self.streamable or False,
+            use_stream_api=self.use_stream_api or False,
+            stream_window_size=self.stream_window_size
+            if self.stream_window_size is not None
+            else 1024 * 1024 * 1024,
```
Comment (clarkzinzow) [inline python/ray/air/config.py:183]: We could automatically set this to e.g. 25% of the cluster's object store capacity.

File: python/ray/air/config.py [hunk 11]
```diff
@@ -146,34 +145,44 @@ class DatasetConfig:
...
+    # When enabled, get_dataset_shard() returns DatasetPipeline instead of Dataset.
+    # The amount of memory to use is controlled by `stream_window_size`.
+    # False by default.
+    use_stream_api: Optional[bool] = None
+
+    # Configure the streaming window size in bytes. A good value is something like
+    # 20% of object store memory. If set to -1, then an infinite window size will be
+    # used (similar to bulk ingest). This only has an effect if use_stream_api is set.
+    # Set to 1.0 GiB by default.
+    stream_window_size: Optional[float] = None
```
Comment (clarkzinzow) [inline python/ray/air/config.py:166]: Once we support streaming aggregations, will this stream window size apply to the fit-side of the preprocessors as well?

**Review by clarkzinzow — APPROVED** [review]
LGTM, great work!

File: python/ray/air/config.py (hunk 9 above)
Comment (ericl) [reply python/ray/air/config.py:170]: Added link.

File: python/ray/air/config.py (hunk 10 above)
Comment (ericl) [reply python/ray/air/config.py:183]: Yeah, I'm a little wary of magic autodetect. Was thinking 1GiB per worker, but even that might be problematic in certain cases.

File: python/ray/air/config.py (hunk 11 above)
Comment (ericl) [reply python/ray/air/config.py:166]: It should, I think.

**Review by richardliaw — APPROVED** [review]
